### PR TITLE
GPT-J Tensor parallel training script and HP updates

### PR DESCRIPTION
**Pull request by anapt**
*Issue #, if available:* N/A

*Description of changes:*
- included script that supports tensor parallelism with the latest Sagemaker Model Parallel release and Hugging Face DLC
- updated existing notebook to use the latest HuggingFace DLC
- moved *.png files to /img
- updated data_pipeline.py to include BertPretrainingDataset  and WikiPretrainingDataset

*Testing done:*
Notebooks execute without errors on us-west-2. 
train_gptj_smp_tensor_parallel_script.py tested by the SMP service team.

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-sagemaker-examples/blob/master/CONTRIBUTING.md) doc and adhered to the example notebook best practices
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-sagemaker-examples/blob/master/README.md)
- [x] I have tested my notebook(s) and ensured it runs end-to-end
- [x] I have linted my notebook(s) and code using `tox -e black-format,black-nb-format`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: training/distributed_training/pytorch/model_parallel/gpt-j/fp16/__init__.py [hunk 1]
```diff
@@ -1,6 +1,5 @@
 # coding=utf-8
-# Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
-# Modifications Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt-j/fp16/__init__.py:3]: Keep the Amazon modifications copyright?

File: training/distributed_training/pytorch/model_parallel/gpt-j/fp16/megatron/clip_grads.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
+# coding=utf-8
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```
[the remark this answers is not in the document]
Comment (anapt) [reply training/distributed_training/pytorch/model_parallel/gpt-j/fp16/megatron/clip_grads.py]: Included proper copyright.

File: training/distributed_training/pytorch/model_parallel/gpt-j/fp16/__init__.py (hunk 1 above)
Comment (anapt) [reply training/distributed_training/pytorch/model_parallel/gpt-j/fp16/__init__.py]: Included proper copyright.

File: training/distributed_training/pytorch/model_parallel/gpt-j/01_train_gptj_smp_notebook.ipynb [hunk 3]
```diff
@@ -2,6 +2,7 @@
  "cells": [
```
Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt-j/01_train_gptj_smp_notebook.ipynb:2]: <p>Is this using a Feature Store variable from another notebook? If so, explicitly mention which notebook and what/why it's doing.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3360/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt-j/01_train_gptj_smp_notebook.ipynb:2]: <pre class="ql-syntax" spellcheck="false">Line #7.            "prescaled_batch": 0,  # was 1
</pre><p>What does it mean that the prescaled batch is 0?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3360/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt-j/01_train_gptj_smp_notebook.ipynb:2]: <pre class="ql-syntax" spellcheck="false">Line #5.    instance_type = "ml.p4d.24xlarge"
</pre><p>Like in the other notebook, explain how customers can get access to this instance type.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3360/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

File: training/distributed_training/pytorch/model_parallel/gpt-j/11_train_gptj_smp_tensor_parallel_notebook.ipynb [hunk 4]
```diff
@@ -0,0 +1,988 @@
+{
```
Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt-j/11_train_gptj_smp_tensor_parallel_notebook.ipynb:1]: <p>Why is us-west-2 required? Customers have region preferences and restrictions for various reasons.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3360/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt-j/11_train_gptj_smp_tensor_parallel_notebook.ipynb:1]: <p>Grammar/branding - consistently use <code>GPT-J</code> rather than <code>GPTJ</code> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3360/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt-j/11_train_gptj_smp_tensor_parallel_notebook.ipynb:1]: <p>Are you able to use the AWS-hosted version at <code>s3://sagemaker-sample-files/datasets/text/SST2/</code>? (train.csv and test.csv files under that prefix)  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3360/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt-j/11_train_gptj_smp_tensor_parallel_notebook.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #34.        #     num_proc=args.preprocessing_num_workers,
</pre><p>Nit: remove commented line?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3360/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt-j/11_train_gptj_smp_tensor_parallel_notebook.ipynb:1]: <p>Nit: is this an extra "store" comment that can be removed? Or does the command print out the stored vars? Also, why is store necessary rather than setting variables?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3360/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt-j/11_train_gptj_smp_tensor_parallel_notebook.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #6.    # training_dataset_location = YOUR_S3_BUCKET/training
</pre><p>Style: please use brackets around any variables that the user would edit, ie <code>&lt;YOUR_S3_BUCKET&gt;</code> (2+ occurrences) </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3360/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt-j/11_train_gptj_smp_tensor_parallel_notebook.ipynb:1]: <p>Handle the <code>else</code> case if the user is running this in another region, so it at least fails elegantly.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3360/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt-j/11_train_gptj_smp_tensor_parallel_notebook.ipynb:1]: <p>Get <code>sagemaker.Session().default_bucket()</code> rather than manually constructing the bucket name? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3360/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt-j/11_train_gptj_smp_tensor_parallel_notebook.ipynb:1]: <p>Add a brief intro/description about FSx.</p><p> </p><p>Branding - change all occurrences of <code>fsx</code>,<code>FSX</code>, etc to <code>FSx</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3360/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='37'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt-j/11_train_gptj_smp_tensor_parallel_notebook.ipynb:1]: <p>What is 13B?</p><p> </p><p>Accessibility - use page title for link text rather than "here" - ie <code>See the instructions at Distributed Training of Mask-RCNN in Amazon SageMaker using FSx...</code></p><p> </p><p>Grammar: <code>Setup</code> -&gt; <code>Set up</code>   </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3360/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='38'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt-j/11_train_gptj_smp_tensor_parallel_notebook.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #8.        file_system_id = "fs-id"
</pre><p>More examples of spots to add brackets for user-specified values, like <code>"&lt;fs-id&gt;"</code>  <span class="ql-cursor">﻿</span></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3360/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='39'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt-j/11_train_gptj_smp_tensor_parallel_notebook.ipynb:1]: <p>Recommended to provide an explanation/link for how to request a quota limit, since both ml.p4d.24xlarge and ml.p3.16xlarge have a default quota of 0 instances.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3360/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='47'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt-j/11_train_gptj_smp_tensor_parallel_notebook.ipynb:1]: <p>What exactly does this shut down? The Studio kernel? Not sure if this is necessary/desirable. You should delete any remaining endpoints, etc, that would incur charge, but I haven't seen this kind of kernel shutdown in a notebook before.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3360/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='60'/>

Comment (anapt) [reply training/distributed_training/pytorch/model_parallel/gpt-j/11_train_gptj_smp_tensor_parallel_notebook.ipynb:1]: updated it to 

> While the above option is easier to setup, using an FSx can be beneficial for performance when dealing with large input sizes and large model sizes. If you are using models with more than 13B parameters, checkpointing should be done using FSx. 

>Please see the instructions at [Distributed Training of Mask-RCNN in Amazon SageMaker using FSx](https://github.com/aws/amazon-sagemaker-examples/blob/master/advanced_functionality/distributed_tensorflow_mask_rcnn/mask-rcnn-scriptmode-fsx.ipynb), to create the FSx lustre filesystem and import the dataset from the S3 bucket to your FSx filesystem. Note that the FSx must be created in a private subnet with internet gateway to ensure that training job has access to the internet.

Comment (anapt) [reply training/distributed_training/pytorch/model_parallel/gpt-j/11_train_gptj_smp_tensor_parallel_notebook.ipynb:1]: removed

Comment (anapt) [reply training/distributed_training/pytorch/model_parallel/gpt-j/11_train_gptj_smp_tensor_parallel_notebook.ipynb:1]: fixed

Comment (anapt) [reply training/distributed_training/pytorch/model_parallel/gpt-j/11_train_gptj_smp_tensor_parallel_notebook.ipynb:1]: fixed

Comment (anapt) [reply training/distributed_training/pytorch/model_parallel/gpt-j/11_train_gptj_smp_tensor_parallel_notebook.ipynb:1]: removed dependency

File: training/distributed_training/pytorch/model_parallel/gpt-j/01_train_gptj_smp_notebook.ipynb (hunk 3 above)
Comment (anapt) [reply training/distributed_training/pytorch/model_parallel/gpt-j/01_train_gptj_smp_notebook.ipynb:2]: added the steps to request a limit increase from AWS Support

Comment (anapt) [reply training/distributed_training/pytorch/model_parallel/gpt-j/01_train_gptj_smp_notebook.ipynb:2]: the prescaled_batch parameter is set to False

Comment (anapt) [reply training/distributed_training/pytorch/model_parallel/gpt-j/01_train_gptj_smp_notebook.ipynb:2]: removed dependency

File: training/distributed_training/pytorch/model_parallel/gpt-j/11_train_gptj_smp_tensor_parallel_notebook.ipynb (hunk 4 above)
Comment (anapt) [reply training/distributed_training/pytorch/model_parallel/gpt-j/11_train_gptj_smp_tensor_parallel_notebook.ipynb:1]: fixed

**Review by jkroll-aws — APPROVED** [review]
